### PR TITLE
 support database parameter DbType and value convertion

### PR DESCRIPTION
**Pull request by hubo0831**

### CODE REVIEW

**Comment by 304NotModified** [comment]
Thanks! Will looks at this in-depth this week/weekend

File: src/NLog/Targets/DatabaseParameterConverter.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    using System.Globalization;
+    using System.Reflection;
+    using NLog.Config;
+    using NLog.Internal;
+    using NLog.Layouts;
+
+    /// <summary>
+    /// Represents SQL command parameter converter.
+    /// </summary>
+    public class DatabaseParameterConverter
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseParameterConverter.cs]: this could be internal?

File: src/NLog/Targets/DatabaseTarget.cs [hunk 2]
```diff
@@ -263,10 +265,30 @@ public DatabaseTarget(string name) : this()
...
+        /// <docgen category='SQL Statement' order='12' />
+        [DefaultValue("DbType")]
+        public string ParameterDbTypePropertyName { get; set; }
+
+        /// <summary>
+        /// Gets or sets converter type of the SQL command parameter value.
+        /// </summary>
+        /// <docgen category='SQL Statement' order='13' />
+        [DefaultValue(typeof(DatabaseParameterConverter))]
+        public Type ParameterConverterType { get; set; }
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseTarget.cs]: this class is not sealed, so I think subclassing is a better option here (with a virtual method). Do you agree?

File: src/NLog/Targets/DatabaseTarget.cs [hunk 3]
```diff
@@ -593,6 +626,33 @@ private void WriteEventToDatabase(LogEventInfo logEvent)
...
+        /// </summary>
+        protected void EnsureResolveParameterInfo(IDbDataParameter p)
+        {
+            if (this.ParameterConverter == null)
+            {
+                lock (this.SyncRoot)
+                {
+                    if (this.ParameterConverter == null)
+                    {
+                        var converter = (DatabaseParameterConverter)Activator.CreateInstance(this.ParameterConverterType);
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseTarget.cs]: as proposed: replace with virtual call

**Comment by 304NotModified** [comment]
thx, added some comments.

File: src/NLog/Targets/DatabaseParameterConverter.cs (hunk 1 above)
Comment (hubo0831) [reply src/NLog/Targets/DatabaseParameterConverter.cs]: No, Others can override this class to support more DbType.

File: src/NLog/Targets/DatabaseTarget.cs (hunk 2 above)
Comment (hubo0831) [reply src/NLog/Targets/DatabaseTarget.cs]: Yes, class 'DatabaseParameterConverter' only implement general methods, others can override this class to support more DbType..

**Comment by hubo0831** [comment]
Sorry, I don't comment instantly, I am on a business trip.

File: src/NLog/Targets/DatabaseTarget.cs (hunk 3 above)
Comment (hubo0831) [reply src/NLog/Targets/DatabaseTarget.cs]: class 'DatabaseTarget' don't need to change, we only inherit from 'DatabaseParameterConverter' to support more DbType.

**Comment by 304NotModified** [comment]
Thanks for the great input!

I have combined and refactor it to #2899 and is expected to be released in NLog 4.6

**Pull request closed without merging**